### PR TITLE
feat: add property based tests for encoding/decoding

### PR DESCRIPTION
**Pull request by jhedev**
You can run the individual test with as many checks as possible, e.g.

```
go test -v -run TestTCPEncodeDecode -rapid.checks=10000 # to run 10k checks
```

Currently, everything passes the checks \o/ I introduced small bugs in the logic
of the methods and quickly the tests failed. So they are indeed working :)

Not sure if we want to keep this which is up to the others. IMO it is a nice way
to generate plenty of testcases.

### CODE REVIEW

**Comment by jhedev** [comment]
Oh and by default this will run 100 checks (as you can see in the CI output)

File: rtuclient_prop_test.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+		if err != nil {
+			t.Fatalf("error while encoding: %+v", err)
+		}
+
+		dpdu, err := packager.Decode(raw)
+		if err != nil {
+			t.Fatalf("error while decoding: %+v", err)
+		}
+
+		if !cmp.Equal(pdu, dpdu) {
```
Comment (gq0) [inline rtuclient_prop_test.go:31]: Do you really need cmp for this basic comparison? I think using std reflect pkg or just directly comparing the `FunctionCode` and `Data` will work. Just trying to minimise dependencies ;)

File: tcpclient_prop_test.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		if err != nil {
+			t.Fatalf("error while encoding: %+v", err)
+		}
+
+		dpdu, err := packager.Decode(raw)
+		if err != nil {
+			t.Fatalf("error while decoding: %+v", err)
+		}
+
+		if !cmp.Equal(pdu, dpdu) {
```
Comment (gq0) [inline tcpclient_prop_test.go:32]: Same as above.

**Review by gq0 — CHANGES_REQUESTED** [review]
Nice. Can you adjust to go.mod change? And I guess cmp is not needed.

File: rtuclient_prop_test.go (hunk 1 above)
Comment (jhedev) [reply rtuclient_prop_test.go:31]: Yeah true. For that simple case we could definitely use `reflect`. However, it is more interesting for the line below `cmp.Diff`. However, I can also change that.

Comment (gq0) [reply rtuclient_prop_test.go:31]: This package is pretty useful. But just for this, it is a little too much 😂 Would be great if you can remove it. Many thanks for the test in general 👍🏻

**Review by gq0 — APPROVED** [review]
I guess the vendoring changed inbetween.

**Review by frzifus — APPROVED** [review]

**Comment by andig** [comment]
What‘s missing to get this PR merged?

**Comment by guelfey** [comment]
@andig basically just adjusting to the switch to go.mod that we did in the meantime

**Comment by andig** [comment]
@jhedev would you kindly follow up? If not I'll try to take a look soonish.

**Comment by jhedev** [comment]
@andig thanks for bring this back to life :) I'm pretty busy with other stuff right now. but I'll try to find time. If you're faster, feel free to take over.

**Comment by andig** [comment]
@jhedev I think all you need is a `git rebase` plus `go mod tidy`. Tried to do that in https://github.com/grid-x/modbus/pull/41 but without much luck in linking that to your pr.

**Comment by jhedev** [comment]
@andig thanks for the hint. I did the rebase and forced pushed again. 

re `cmp`. I'm not only using the `Compare` but also the `Diff` method. Happy to replace it but I still think it is a useful dependency.

**Comment by andig** [comment]
> Happy to replace it but I still think it is a useful dependency.

I wouldn't mind, especially as it's test-only.

@guelfey good to go?

**Comment by andig** [comment]
@jhedev if `go-cmp` remains a blocker you could use `reflect.DeepEqual` and replace the diff by "expected/got". It's a small helper during testing when an error actually occurs, but nothing absolutely required imho.

**Review by guelfey — APPROVED** [review]

**Pull request merged**